From 2b1a636158aa201fef54bd7e59a4c10b795c45a4 Mon Sep 17 00:00:00 2001
From: Yuvasree Sivasankaran <quic_ysivasan@quicinc.com>
Date: Fri, 29 Dec 2023 12:56:18 +0530
Subject: [PATCH] wifi: ath12k: Fix kernel bug during pktlog dump collection

During pktlog dump collection to user space, a kernel bug occurs.
In ath12k_pktlog_read, copy_to_user has sleep function and this api
is called inside pktlog info lock which results in kernel bug.
Unlock pktlog info once it is not needed eventually copy_to_user comes
out of lock.

Below the call stack occurs during pktlog dump

BUG: sleeping function called from invalid context at include/linux/uaccess.h:143
 in_atomic(): 1, irqs_disabled(): 0, non_block: 0, pid: 5295, name: cp
 preempt_count: 201, expected: 0
 RCU nest depth: 0, expected: 0
 CPU: 3 PID: 5295 Comm: cp Not tainted 6.1.31 #0
 Hardware name: Qualcomm Technologies, Inc. IPQ9574/RDP449/AP-AL02-C6 (DT)
 Call trace:
  dump_backtrace+0xd4/0xfc
  show_stack+0x14/0x1c
  dump_stack_lvl+0x70/0x8c
  dump_stack+0x14/0x30
  __might_resched+0x13c/0x14c
  __might_sleep+0x60/0x6c
  __might_fault+0x18/0x24
  ath12k_coredump_m3_dump+0x574/0xea8 [ath12k]
  full_proxy_read+0x54/0x80
  vfs_read+0xa8/0x1fc
  ksys_read+0x70/0xe4
  __arm64_sys_read+0x14/0x1c
  el0_svc_common.constprop.0+0xc0/0x130
  do_el0_svc+0x18/0x20
  el0_svc+0x14/0x3c
  el0t_64_sync_handler+0xa8/0x134
  el0t_64_sync+0x14c/0x150

Signed-off-by: Yuvasree Sivasankaran <quic_ysivasan@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/pktlog.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/drivers/net/wireless/ath/ath12k/pktlog.c b/drivers/net/wireless/ath/ath12k/pktlog.c
index 53e1b58..28bec4a 100644
--- a/drivers/net/wireless/ath/ath12k/pktlog.c
+++ b/drivers/net/wireless/ath/ath12k/pktlog.c
@@ -276,6 +276,8 @@ static ssize_t ath12k_pktlog_read(struct file *file, char __user *userbuf,
 
 	} while (cur_rd_offset != log_buf->wr_offset);
 
+	spin_unlock_bh(&info->lock);
+
 	ppos_data = *ppos + ret_val - bufhdr_size + start_offset;
 
 	if (fold_offset == -1) {
@@ -325,7 +327,6 @@ static ssize_t ath12k_pktlog_read(struct file *file, char __user *userbuf,
 read_done:
 	*ppos += ret_val;
 out:
-	spin_unlock_bh(&info->lock);
 	return ret_val;
 }
 
-- 
2.34.1

